File: apps/backend/internal/recipe/core/models/recipe.go
package models

import "time"

type Recipe struct {
	ID          string
	Title       string
	Description *string
	PrepTime    *uint32
	Servings    *uint32
	IsPublished bool
	CoverImage  *string
	CreatedAt   time.Time
	UpdatedAt   time.Time
	Nutritions  []Nutrition
}

type RecipeStep struct {
	ID          uint32
	RecipeID    string
	StepNumber  uint32
	Description string
}

type RecipeIngredient struct {
	RecipeID      string
	IngredientID  uint32
	MeasurementID *uint32
	Amount        uint32
	Ingredient    string
	Measurement   *string
}

type Nutrition struct {
	ID       uint32
	RecipeID string
	Name     string
	Unit     string
	Value    uint32
}

type CreateRecipeInput struct {
	Title       string
	Description *string
	PrepTime    *uint32
	Servings    *uint32
	IsPublished *bool
}

type UpdateRecipeInput struct {
	Title       *string
	Description *string
	PrepTime    *uint32
	Servings    *uint32
	IsPublished *bool
}

type ListRecipesFilter struct {
	Page        uint32
	Limit       uint32
	IsPublished *bool
}

type CreateRecipeStepInput struct {
	Order       uint32
	Description string
}

type UpdateRecipeStepInput struct {
	Description string
}

type ReorderRecipeStepInput struct {
	OrderFrom uint32
	OrderTo   uint32
}

type CreateRecipeIngredientInput struct {
	IngredientID  uint32
	MeasurementID *uint32
	Amount        uint32
}

type UpdateRecipeIngredientInput struct {
	MeasurementID *uint32
	Amount        uint32
}

type UpdateNutritionInput struct {
	Name  string
	Unit  string
	Value uint32
}

File: packages/types/src/recipe.ts
export type RecipeIngredient = {
  ingredientId: number;
  amount: number;
  measurementId: number | null;
  recipeId: string;
  ingredient: string;
  measurement: string | null;
};

export const nutritions = [
  "carbohydrates",
  "energy",
  "fat",
  "fibers",
  "protein",
  "saturatedFat",
  "sodium",
  "sugar",
] as const;

export type Nutrition = (typeof nutritions)[number];

export const nutritionOrder: Nutrition[] = [
  "energy",
  "carbohydrates",
  "sugar",
  "sodium",
  "protein",
  "fat",
  "saturatedFat",
  "fibers",
];

export const nutritionUnits = ["g", "mg", "kcal"] as const;

export type NutritionUnit = (typeof nutritionUnits)[number];

export type RecipeNutrition = {
  value: number;
  unit: string;
  id: number;
  name: string;
  recipeId: string;
};

export type RecipeStep = {
  description: string;
  id: number;
  recipeId: string;
  stepNumber: number;
};

export type Recipe = {
  id: string;
  title: string;
  description: string | null;
  prepTime: number | null;
  servings: number | null;
  isPublished: boolean;
  coverImageUrl: string | null;
  createdAt: Date;
  updatedAt: Date;
  nutrition: Partial<Record<Nutrition, RecipeNutrition>>;
};
